### PR TITLE
Added support for webauth env variable, re-arranged css declarations

### PR DESCRIPTION
**Pull request by jgreben**

### CODE REVIEW

**Comment by jgreben** [comment]
@dlrueda want to merge this? More basic things. I rearranged the CSS so it would favor the fonts @jvine will want, and added in basic support for webauth.

File: app/controllers/authentication_controller.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+###
+#  Controller for handling login/logout and redirecting with flash messages.
+###
+class AuthenticationController < ApplicationController
+  def login
+    session[:sunet] = request.env['REMOTE_USER'] || ENV['REMOTE_USER']
```
Comment (jkeck) [inline app/controllers/authentication_controller.rb]: This feels potentially problematic.  We typically don't store data like SUNet IDs in session data.

A more idiomatic approach might be to have a user class that models the attributes about a user (probably just SUNet ID right now).  That object can be instantiated in the controller context (e.g. `ApplicationController`) and be made available to all the views (typically under `current_user`).

This also follows the pattern of libraries like CanCan and devise, which would allow you to plug them in more easily of you so choose.

File: app/controllers/authentication_controller.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    flash[:success] = 'You have been successfully logged in.'
+    redirect_to params[:referrer] || :back
+  end
+
+  def logout
+    session[:sunet] = nil
+    redirect_to 'http://weblogin.stanford.edu/logout'
```
Comment (jkeck) [inline app/controllers/authentication_controller.rb]: Hm, I haven't seen this pattern for logging out in a rails app.  Typically, we configure a local endpoint (in apache) that logs the user out.  Perhaps @eefahy or @tallenaz can help get apache configured to log users out (which might also be easier to migrate once WebAuth is EOLed).

File: app/controllers/authentication_controller.rb (hunk 1 above)
Comment (jgreben) [reply app/controllers/authentication_controller.rb]: 👍

File: app/controllers/authentication_controller.rb (hunk 2 above)
Comment (jgreben) [reply app/controllers/authentication_controller.rb]: 👍

**Comment by jgreben** [comment]
@jkeck made the changes you suggested (used `sul-requests` as a reference for the User and CurrentUser models). Looked at the apache config on the `libsys-webforms` box and now I'm grokking on what you mentioned about webauth logout. Thanks.

**Comment by jgreben** [comment]
@jkeck made the changes you suggested (used `sul-requests` as a reference for the User and CurrentUser models). Looked at the apache config on the `libsys-webforms` box and now I'm grokking on what you mentioned about webauth logout. Thanks.

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+###
+#  User class for authenticating via WebAuth
+###
+class User < ActiveRecord::Base
```
Comment (jkeck) [inline app/models/user.rb]: Just out of curiousity, is there a requirement to store these users?  I'm not familiar enough w/ the whole app to know we'll need user records persisted in the DB or not. If not we could probably get away with a plain ruby object that doesn't require any persistence.

If you know that you def. need to persist user data and associate user records with other objects, then 👍

File: app/models/current_user.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  end
+
+  private
+
+  def webauth_user
+    User.find_or_create_by(webauth: user_id)
+  end
+
+  def anonymous_user
+    User.new(ip_address: request.remote_ip)
```
Comment (jkeck) [inline app/models/current_user.rb]: Do we currently have any use for this up_address passing for this anonymous user object?

**Review by jkeck — COMMENTED** [review]
Just a few more comments about the User class.

This is starting to look good!

File: app/models/user.rb (hunk 3 above)
Comment (jgreben) [reply app/models/user.rb]: There is not a defined use case that would require storage of users at this time, so let me try switching this away from being an ActiveRecord object. (And for that matter, the database in production will need to be a Symphony Oracle table for other reasons that we can discuss later.)

**Comment by jgreben** [comment]
@jkeck I made the changes as suggested, which simplified some of the tests, and I dropped the Users table. Let me know what you think.

File: db/migrate/20170503212744_drop_user_table.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class DropUserTable < ActiveRecord::Migration
```
Comment (jkeck) [inline db/migrate/20170503212744_drop_user_table.rb]: Since this PR was never merged, you can just remove all these migrations (and that way you don't have to create a table just to remove it in a later migration).

File: app/views/shared/_top_navbar.html.erb [hunk 6]
```diff
@@ -7,7 +7,13 @@
     <% end %>
 
     <div class="header-links" id="login_id">
-        <%= link_to "Feedback", "http://searchworks.stanford.edu/feedback", target: '_blank', rel: 'noopener noreferrer'  %>
+      <% if current_user != nil && current_user.webauth_user? %>
```
Comment (jkeck) [inline app/views/shared/_top_navbar.html.erb]: Ah, I see you have to do this because we dropped the Null/Anonymous user all-together.  That's cool, although depending on how much we deal w/ the user object we may want to build that back in down the road.

In the meantime, this could be more sussinctly written as `if current_user && current_user.webauth_user?`.

If you find that you are doing the whole `if current_user && ...` frequently around the app, it may be a good idea to re-implement some sort of Null/Anonymous user object.

**Review by jkeck — COMMENTED** [review]
Okay, last comments, I swear. 😂

File: app/views/shared/_top_navbar.html.erb (hunk 6 above)
Comment (jgreben) [reply app/views/shared/_top_navbar.html.erb]: 👍 Ok cool, I'll keep an eye out for that! Made the change as suggested.

File: db/migrate/20170503212744_drop_user_table.rb (hunk 5 above)
Comment (jgreben) [reply db/migrate/20170503212744_drop_user_table.rb]: 👍

**Comment by jgreben** [comment]
Ok @jkeck, ready to merge? 🙏

File: .rubocop.yml [hunk 7]
```diff
@@ -33,6 +33,11 @@ Metrics/MethodLength:
 Metrics/ModuleLength:
   Max: 120
 
+# Disabled since it only applies to Rails 5. Reenable when we migrate our apps to Rails 5.
```
Comment (cbeer) [inline .rubocop.yml]: Is there any reason not to move to Rails 5 now? Rails 4.2 is more-or-less end-of-life right now (major security patches only), so....

File: app/models/user.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+###
+class User
+  attr_accessor :webauth
+
+  def initialize(webauth)
+    @webauth = webauth
+  end
+
+  def webauth_user?
+    @webauth.present?
```
Comment (cbeer) [inline app/models/user.rb]: You defined an `attr_accessor :webauth` above, but aren't using it?

Comment (jgreben) [reply app/models/user.rb]: It seems to be needed by `<%= link_to "#{current_user.webauth}: Logout", logout_path(referrer: root_path) %>` in  _top_navbar.html.erb and tests

File: .rubocop.yml (hunk 7 above)
Comment (jgreben) [reply .rubocop.yml]: Do we prefer rails 5.0 or bleeding edge 5.1?

File: app/models/user.rb (hunk 8 above)
Comment (cbeer) [reply app/models/user.rb]: That's fine -- then I guess I'm wondering why you're using `@webauth` in `User` instead of the `webauth` accessor you already have?

File: .rubocop.yml (hunk 7 above)
Comment (cbeer) [reply .rubocop.yml]: Either is fine.

File: app/models/user.rb (hunk 8 above)
Comment (jgreben) [reply app/models/user.rb]: If I don't use `@webauth` then webauth ends up being `nil`. I can't seem to get it to work any other way...

Comment (jkeck) [reply app/models/user.rb]: @jgreben Just so I'm clear, you're saying that changing line 12 in this file from `@webauth.present?` to `webauth.present?` doesn't work?

Comment (jgreben) [reply app/models/user.rb]: 😖 yes, that works. Sorry, I got confused and though Chris was talking about the initialize method.

File: .rubocop.yml (hunk 7 above)
Comment (jgreben) [reply .rubocop.yml]: I upgraded the app to rails 5.0. Everything seems to work, except when I run `rake` I am getting:
```
DEPRECATION WARNING: `redirect_to :back` is deprecated and will be removed from Rails 5.1. Please use `redirect_back(fallback_location: fallback_location)` where `fallback_location` represents the location to use if the request has no HTTP referer information. (called from login at /Users/jgreben/Projects/MyAccount/app/controllers/authentication_controller.rb:7)
```
 Except when I try doing that I get:
```
undefined method `redirect_back' for #<RSpec::ExampleGroups...
```

My bundle update says I am using rspec-core 3.5.4 and rspec-expectations 3.5.0

Comment (jgreben) [reply .rubocop.yml]: Ok, never-mind, `redirect_back` was supposed to go in the controller, not the controller spec. This is working now.

**Comment by jgreben** [comment]
@jkeck, @cbeer is this OK to merge now?

File: Gemfile [hunk 9]
```diff
@@ -1,8 +1,7 @@
 source 'https://rubygems.org'
 
-
 # Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
-gem 'rails', '4.2.7.1'
+gem 'rails', '5.0'
```
Comment (jkeck) [inline Gemfile]: This seems a bit too strict.  I think if you use something along the lines of `~> '5.0.0'` you'll be able to upgrade to different patch versions of rails (for instance, 5.0.2 is what we should probably be using, but the way this is specified we're only going to get 5.0.0 exactly).

Luckily you've done the heavy lifting updating to rails 5, so that's a simple change.

Comment (jgreben) [reply Gemfile]: 👍 Thanks! I made the change and did a bundle update. Now we have:
```
Installing activemodel 5.0.2 (was 5.0.0)
...
Installing activejob 5.0.2 (was 5.0.0)
Installing activerecord 5.0.2 (was 5.0.0)
Installing actionview 5.0.2 (was 5.0.0)
Installing actionpack 5.0.2 (was 5.0.0)
Installing actioncable 5.0.2 (was 5.0.0)
Installing actionmailer 5.0.2 (was 5.0.0)
Installing railties 5.0.2 (was 5.0.0)
...
Installing rails 5.0.2 (was 5.0.0)
```

**Comment by jgreben** [comment]
@jkeck ready to merge?

**Review by jkeck — APPROVED** [review]
